[12.x] Const to enum

shaedrich opened the pull request:
Maybe, it's time for Laravel to not only allow custom enums but also to have it's own

* Moved enumerable class constants to separate enums
* Tried to stay backward compatible by
    * Deprecating the constants instead of removing them
    * Allowing to pass both the enum and the underlying value to a function and wrap the argument in an `enum_value()` call

shaedrich committed:
const to enum

shaedrich committed:
Fix syntax error

shaedrich committed:
Fix StyleCI

shaedrich committed:
More linting fixes

shaedrich committed:
Fix Typo

shaedrich committed:
Fix typo

shaedrich committed:
Add missing input

shaedrich committed:
stringify enums

shaedrich committed:
Fix typo

shaedrich committed:
enum_value

shaedrich committed:
Fix test

shaedrich committed:
Another enum_value

shaedrich committed:
FIx StyleCI

shaedrich committed:
Fix syntax error

shaedrich committed:
Add missing import

shaedrich committed:
Fix StyleCI

shaedrich committed:
Only wrap in enum if it is none

tontonsb commented on src/Illuminate/Console/Scheduling/ManagesFrequencies.php:
```diff
@@ -2,20 +2,23 @@
      * The Cron expression representing the event's frequency.
      *
-     * @param  string  $expression
+     * @param  string|\Illuminate\Console\Enums\ScheduleOn  $expression
```
Is this method even supposed to accept single integers like `->cron(3)`? I thought it's intended for those cron strings like `0 15 10 ? * 6L 2002-2005`.

shaedrich replied:
I decided to add validation in a separate PR: #53971

tontonsb commented:
Regarding casing... AFAIK the vast majority of devs use editors with syntax highlighting. At least Notepad++ or Vim. So there's no need to set apart constants/enum cases by uppercasing them. Although the uppercase practice is too engrained for constants, we might consider not continuing it for enums. Wouldn't `PasswordStatus::passwordReset` or `PasswordStatus::PasswordReset` be nicer?

Overall enums are intended for finite sets of values you're willing to accept and handle, they make an unextensible list of values. If the values are just "some values" or even worse "some of the values", constants or plain classes should be used instead. Enums are not a different syntax for constants, they are types. I.e. `public function myfun(HandledCase $case)` should be used if you would otherwise do this:

```php
public function myfun(string $case)
{
    if (!in_array($case, static::handledCases))
        throw new \DomainException("Invalid case value $case here.");

    // ....
}
```

See also https://peakd.com/hive-168588/@crell/on-the-use-of-enums

Do we want to limit any of these into a final, non-extensible list of cases? Even for weekdays (which afaik make a finite list per se) we're talking about the weekday portion of a cron string which should accept `0`, `7`, `2-6`, `tue-fri`, `4#2`, `5L`. I think in all of those cases constants were the appropriate choice as aliases/sugar over the most common values.

taylorotwell commented:
Agree with casing comment. I would only change Password reset and Day scheduling stuff to Enums. The worker stuff shouldn't be changed imo. It's not user facing.

shaedrich commented:
@taylorotwell Thanks for the clear instructions—I'll change that 👍🏻

shaedrich committed:
Remove worker-related enum

as per https://github.com/laravel/framework/pull/53961#issuecomment-2552043551

Co-authored by: Taylor Otwell <taylor@laravel.com>

shaedrich committed:
Change casing for enum cases

Co-authored-by: Džuris <juris@glaive.pro>

shaedrich committed:
Remove unused function

Follow-up to a8eca1b

Co-authored-by: Taylor Otwell <taylor@laravel.com>

shaedrich committed:
Revert to promoted property

Follow-up to a8eca1b

shaedrich committed:
Remove enum_value to the proper location

shaedrich committed:
Cron expression is not an enum

Follow-up to f49854a

Co-Authored-By: rodrigopedra <rodrigo.pedra@gmail.com>
Co-authored-by: Džuris <juris@glaive.pro>

tontonsb commented on src/Illuminate/Console/Scheduling/Schedule.php:
```diff
@@ -28,19 +29,26 @@ class Schedule
 
-    const SATURDAY = 6;
+    /** @deprecated Use \Illuminate\Console\Enums\ScheduleOn::SATURDAY instead */
+    const SATURDAY = ScheduleOn::SATURDAY;
```
I like your deprecation strategy.

tontonsb commented on src/Illuminate/Queue/Events/WorkerStopping.php:
```diff
@@ -15,5 +15,6 @@ public function __construct(
         public $status = 0,
         public $workerOptions = null
     ) {
+        $this->status = $status;
```
Did you intend this change or is that a leftover from a previous iteration of this PR?

shaedrich replied:
[Good catch 👍🏻 It's a leftover, I missed](https://github.com/laravel/framework/pull/53961#discussion_r1891250982)

tontonsb commented on src/Illuminate/Queue/Worker.php:
```diff
@@ -304,12 +304,12 @@ protected function pauseWorker(WorkerOptions $options, $lastRestart)
+            $this->queueShouldRestart($lastRestart) => self::EXIT_SUCCESS,
+            $options->stopWhenEmpty && is_null($job) => self::EXIT_SUCCESS,
+            $options->maxTime && hrtime(true) / 1e9 - $startTime >= $options->maxTime => self::EXIT_SUCCESS,
+            $options->maxJobs && $jobsProcessed >= $options->maxJobs => self::EXIT_SUCCESS,
```
Could this also be a leftover? Or is the change from `static` to `self` intentionally left as a part of this PR?

rodrigopedra replied:
While reviewing this file, I also felt compelled to say this change was unnecessary.

But after some thinking, and considering your previous remark regarding when enums should be used versus when constants should be used, I thought that, even if the class isn't final, there is no point on reading these constants from subclasses, as they are basically hard-coded exit values from processes, and should not change.

Using `self::` instead of `static::`, improves -- IMO -- intention declaration and correctness in this case.

rodrigopedra commented on src/Illuminate/Console/Enums/ScheduleOn.php:
```diff
@@ -0,0 +1,20 @@
+
+enum ScheduleOn: int
+{
+    case SUNDAY = 0;
```
Per the comments regarding casing, I believe we could use `PascalCase` here too

rodrigopedra commented on src/Illuminate/Console/Scheduling/Schedule.php:
```diff
@@ -28,19 +29,26 @@ class Schedule
 
-    const SUNDAY = 0;
+    /** @deprecated Use \Illuminate\Console\Enums\ScheduleOn::SUNDAY instead */
+    const SUNDAY = ScheduleOn::SUNDAY;
```
As this is scheduled for the next version, I would take the opportunity to move completely to the new approach, and document these changes.

But this is better to check with one of the maintainers. Maybe it is better to keep those for a smooth transition

rodrigopedra commented on src/Illuminate/Queue/Events/WorkerStopping.php:
```diff
@@ -15,5 +15,6 @@ public function __construct(
         public $status = 0,
         public $workerOptions = null
     ) {
+        $this->status = $status;
```
Why do we need this? Isn't this property a promoted property already?

shaedrich replied:
Good catch 👍🏻 It's a leftover, I missed

tontonsb reviewed (commented):
I'm still thinking about the finiteness of enums. What if I had overriden `sendResetLink` to also return some other case like `'passwords.awaiting admin approval­'`.?

Sure, the return type is only a comment, so technically we'd still allow that. But conceptually/semantically enum return type means that no other cases are allowed. Is the set of password reset statuses something that should be forbidden to extend?

shaedrich committed:
Also make ScheduleOn cases PascalCase

Follow-up to 935efcc

Co-Authored-By: rodrigopedra <rodrigo.pedra@gmail.com>
Co-authored-by: Džuris <juris@glaive.pro>

shaedrich committed:
More casing

Follow-up to f44b034

shaedrich committed:
Remove manually setting argument that has been reverted to a promited property

Follow-up to 7f0bda7

Co-Authored-By: rodrigopedra <rodrigo.pedra@gmail.com>
Co-authored-by: Džuris <juris@glaive.pro>

shaedrich commented:
> Even for weekdays (which afaik make a finite list per se) we're talking about the weekday portion of a cron string which should accept `0`, `7`, `2-6`, `tue-fri`, `4#2`, `5L`. I think in all of those cases constants were the appropriate choice as aliases/sugar over the most common values.

As far as I can see, they are not type-hinted anywhere in a way that would restrict you to pass anything else. They currently only appear to my knowledge in places where these are passed directly. You can still pass anything you like to `ManageFrequencies::days()`

shaedrich committed:
Deprecation messages and correct PHPDoc types

Follow-up to 3a251cf9546bf3d050904aba200ee54e1aa9ee3c

shaedrich committed:
Improve type-hint for days

shaedrich committed:
More PHPDocs

shaedrich committed:
Fix StyleCI

https://github.styleci.io/analyses/KoMvmA

shaedrich committed:
More StyleCI

https://github.styleci.io/analyses/xgyApB

shaedrich committed:
Remove accidentally added spacing

https://github.styleci.io/analyses/xgyApB

shaedrich committed:
StyleCI leftovers

Follow-up to 0e26932 and f384afdb569c5fff0c606c98a76e5682ecb9cb9f

https://github.styleci.io/analyses/O3wGZ0

Co-authored-by: StyleCI Bot <bot@styleci.io>

tontonsb commented:
> As far as I can see, they are not type-hinted anywhere in a way that would restrict you to pass anything else.

@shaedrich yeah, but type-hinting in a way that would restrict you to pass anything else is the reason why enums exist. They are types, not a syntax for magic values (that's the job of constants). If you're not going to limit the list of values, enums are not the right tool.

shaedrich commented:
@tontonsb I think, we are discussing two separate things here. That might be why we talked slightly past each other.
* Replacing (non-finite) constants with (finite) enums
* Using enums as (restricting/enforcing) type hints (doesn't happen in this PR)

So, the former is something, we probably have to wait for Taylor's opinion.

shaedrich committed:
Months are 1-indexed

tontonsb commented:
@shaedrich I think I understand your point, I'm just trying to say that the former is a usecase for constants not enums. Only the latter is for enums.

If you don't want a restricting type, but want the list to be finite, that is still a job for constants. Here's an example from the [same article](https://peakd.com/hive-168588/@crell/on-the-use-of-enums) I linked earlier:

```php
final class Role
{
   public const Admin = 'admin';
   public const Editor = 'editor;
   public const User = 'user';
   public const AnonymousCoward = 'anonymous';
}
```

You probably know that already, but for future readers — the author of that article is not some random enum hater. He [introduced enums in PHP](https://wiki.php.net/rfc/enumerations) himself.

shaedrich committed:
Add type-hints to respective facade

shaedrich committed:
Add range type-hint for seconds

shaedrich committed:
Add array type-hint

shaedrich committed:
Fix typo

shaedrich commented:
> @shaedrich I think I understand your point, I'm just trying to say that the former is a usecase for constants not enums. Only the latter is for enums.

@tontonsb I agree: Constants are no enums and both have their respective place. However, the question remains: _Should_ this even be a non-finite list in the first place?

taylorotwell committed:
formatting

taylorotwell commented:
Renamed `ScheduleOn` to `Day` and `PasswordStatus` to `PasswordResetResult`.

taylorotwell committed:
Update PasswordResetResult.php

taylorotwell committed:
still use strings for passwords

taylorotwell commented:
Actually keeping `PasswordResetResult` as constants on a class. I like the camel casing and better naming, but don't want the breaking change of switching to an enum, which is not a string.

taylorotwell committed:
cant depend on auth component

taylorotwell committed:
dont use enums

taylorotwell commented:
Eh, imo, this thing is getting a bit overcooked. I added camel case constants for password reset on 11.x.

If you want to separately PR a simple Day enum go for it. I don't see why it couldn't go to 11.x.

shaedrich commented:
see #53983
